File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/DownloadJSONService.java
package com.brentpanther.bitcoinwidget;

import android.app.IntentService;
import android.content.Intent;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;

import java.io.FileOutputStream;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;


public class DownloadJSONService extends IntentService {

    private static final String LAST_MODIFIED = "last_modified";
    public static final String CURRENCY_FILE_NAME = "coins.json";
    public static final String JSON_DOWNLOADED_ACTION = "json-downloaded";
    private static final String TAG = DownloadJSONService.class.getSimpleName();
    static boolean downloaded = false;

    public DownloadJSONService() {
        super("Download JSON");
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        try {
            downloadJSON();
        } catch (IOException e) {
            Log.e(TAG, "Error downloading JSON.", e);
        }
        downloaded = true;
        LocalBroadcastManager.getInstance(this).sendBroadcast(new Intent(JSON_DOWNLOADED_ACTION));
    }

    private void downloadJSON() throws IOException {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
        String lastModified = prefs.getString(LAST_MODIFIED, getString(R.string.json_last_modified));
        String url = getString(R.string.json_url);
        OkHttpClient client = new OkHttpClient.Builder()
                .followRedirects(true)
                .followSslRedirects(true)
                .readTimeout(5, TimeUnit.SECONDS)
                .connectTimeout(5, TimeUnit.SECONDS)
                .build();
        Request request = new Request.Builder()
                .addHeader("If-Modified-Since", lastModified)
                .url(url)
                .build();

        Response response = client.newCall(request).execute();
        switch (response.code()) {
            case 304:
                Log.d(TAG, "No changes found in JSON file.");
                break;
            default:
                Log.d(TAG, "Retrieved status code: " + response.code());
                break;
            case 200:
                Log.d(TAG, "Updated JSON file found.");
                prefs.edit().putString(LAST_MODIFIED, response.header("Last-Modified")).apply();
                byte[] json = response.body().bytes();
                FileOutputStream os = openFileOutput(CURRENCY_FILE_NAME, MODE_PRIVATE);
                os.write(json);
                os.close();
        }
    }

}

File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/DownloadJSONService.kt
package com.brentpanther.bitcoinwidget

import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.annotation.Nullable
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.preference.PreferenceManager
import okhttp3.ConnectionPool
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.util.concurrent.TimeUnit


class DownloadJSONService : IntentService("Download JSON") {

    override fun onHandleIntent(@Nullable intent: Intent?) {
        try {
            downloadJSON()
        } catch (e: IOException) {
            Log.e(TAG, "Error downloading JSON.", e)
        }

        downloaded = true
        LocalBroadcastManager.getInstance(this).sendBroadcast(Intent(JSON_DOWNLOADED_ACTION))
    }

    @Throws(IOException::class)
    private fun downloadJSON() {
        val prefs = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        val lastModified = prefs.getString(LAST_MODIFIED, getString(R.string.json_last_modified))
        val url = getString(R.string.json_url)
        val client = OkHttpClient.Builder()
                .followRedirects(true)
                .followSslRedirects(true)
                .readTimeout(5, TimeUnit.SECONDS)
                .connectTimeout(5, TimeUnit.SECONDS)
                .retryOnConnectionFailure(true)
                .connectionPool(ConnectionPool(0, 1, TimeUnit.NANOSECONDS))
                .build()
        val request = Request.Builder()
                .addHeader("If-Modified-Since", lastModified!!)
                .url(url)
                .build()

        val response = client.newCall(request).execute()
        when (response.code) {
            304 -> Log.d(TAG, "No changes found in JSON file.")
            200 -> {
                Log.d(TAG, "Updated JSON file found.")
                prefs.edit().putString(LAST_MODIFIED, response.header("Last-Modified")).apply()
                val json = response.body!!.bytes()
                val os = openFileOutput(CURRENCY_FILE_NAME, Context.MODE_PRIVATE)
                os.write(json)
                os.close()
            }
            else -> Log.d(TAG, "Retrieved status code: " + response.code)
        }
    }

    companion object {

        private const val LAST_MODIFIED = "last_modified"
        const val CURRENCY_FILE_NAME = "coins.json"
        const val JSON_DOWNLOADED_ACTION = "json-downloaded"
        private val TAG = DownloadJSONService::class.java.simpleName
        internal var downloaded = false
    }

}
